File: java/parser/SyntaxError.java
/*
 * Copyright (C) 2020 Grakn Labs
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package graql.lang.parser;

import java.util.Objects;

import static graql.lang.common.exception.ErrorMessage.SYNTAX_ERROR_DETAILED;
import static graql.lang.common.exception.ErrorMessage.SYNTAX_ERROR_NO_DETAILS;

public class SyntaxError {

    private final String queryLine;
    private final int line;
    private final int charPositionInLine;
    private final String msg;
    private final int hash;

    public SyntaxError(String queryLine, int line, int charPositionInLine, String msg) {
        if (msg == null) throw new NullPointerException("Null msg");
        this.queryLine = queryLine;
        this.line = line;
        this.charPositionInLine = charPositionInLine;
        this.msg = msg;
        this.hash = Objects.hash(this.queryLine, this.line, this.charPositionInLine, this.msg);
    }

    private String spaces(int len) {
        final char ch = ' ';
        final char[] output = new char[len];
        for (int i = len - 1; i >= 0; i--) {
            output[i] = ch;
        }
        return new String(output);
    }

    @Override
    public String toString() {
        if (queryLine == null) {
            return SYNTAX_ERROR_NO_DETAILS.message(line, msg);
        } else {
            // Error message appearance:
            //
            // syntax error at line 1:
            // match $
            //       ^
            // blah blah antlr blah
            final String pointer = spaces(charPositionInLine) + "^";
            return SYNTAX_ERROR_DETAILED.message(line, queryLine, pointer, msg);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (o == this) {
            return true;
        }
        if (o instanceof SyntaxError) {
            final SyntaxError that = (SyntaxError) o;
            return (Objects.equals(this.queryLine, that.queryLine) &&
                    this.line == that.line &&
                    this.charPositionInLine == that.charPositionInLine &&
                    this.msg.equals(that.msg));
        }
        return false;
    }

    @Override
    public int hashCode() {
        return hash;
    }
}

File: rust/parser/syntax_error.rs
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

use std::fmt;

use crate::common::error::{ErrorMessage, SYNTAX_ERROR_DETAILED, SYNTAX_ERROR_NO_DETAILS};

#[derive(Clone, Debug)]
pub struct SyntaxError {
    pub query_line: Option<String>,
    pub line: usize,
    pub char_position_in_line: usize,
    pub message: String,
}

impl From<SyntaxError> for ErrorMessage {
    fn from(syntax_error: SyntaxError) -> Self {
        if let Some(query_line) = &syntax_error.query_line {
            // Error message appearance:
            //
            // syntax error at line 1:
            // match $
            //       ^
            // blah blah antlr blah
            SYNTAX_ERROR_DETAILED.format(&[
                syntax_error.line.to_string().as_str(),
                query_line,
                &(" ".repeat(syntax_error.char_position_in_line) + "^"),
                &syntax_error.message,
            ])
        } else {
            SYNTAX_ERROR_NO_DETAILS
                .format(&[syntax_error.line.to_string().as_str(), &syntax_error.message])
        }
    }
}

impl fmt::Display for SyntaxError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", ErrorMessage::from(self.clone()).message)
    }
}
